### DISCUSSION TITLE
Is it possible to convert an effectful function into a `IO` function to be stored in a data structure?

### DISCUSSION THREAD

**Discussion by newhoggy** [haskell-effectful/effectful#275] [Q&A] [answered]
I need to store an `IO` function that logs in a data type for my logging library to use and it must be derivable from an effectful function:

https://github.com/haskell-works/effectful-zoo/pull/1/files#r1848260662

Is that possible?

Either way, is it the best way for me to do what I'm trying to do?

**Comment by ocharles** [comment]
You want to use `withEffToIO`. As to which unlifting strategy, it depends a bit more on how you plan to call these `IO` actions. Will they be called from the same thread, or different threads?

**Reply by newhoggy** [reply]
It will initially be the same thread, however the computation could spawn threads itself, and I'd like it to continue to work.

**Reply by ocharles** [reply]
Ok, then you want to use either the `ConcUnlift` or `ForkUnlift` strategies. You would do something like:

```haskell
withEffToIO (ConcUnlift .. ..) \runInIO -> runInIO do
  -- Your whole app runs here
  foo Foo{ doThing = runInIO thing }
```

or

```haskell
runInIO <- withEffToIO ForkUnlift return
...
foo Foo{ doThing = runInIO thing }
```

I think those are the correct way to use these!

**Reply by arybczak** [reply, marked as answer]
If it's for logging in multiple threads, `ConcUnlift Persistent Unlimited` seems like the best option, but generally I wouldn't do it this way and use existing library, either `log-effectful` or `co-log-effectful`.

> ForkUnlift

It's `SeqForkUnlift` and thus only for a single thread.

**Comment by newhoggy** [comment]
Thanks!
